File: app/src/main/java/com/example/mvvm_sample/viewmodel/SampleAndroidViewModelFactory.java
package com.example.mvvm_sample.viewmodel;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.example.mvvm_sample.model.ISampleModel;

import java.lang.reflect.InvocationTargetException;

public class SampleAndroidViewModelFactory implements ViewModelProvider.Factory {

    private Application application;
    private ISampleModel sampleModel;

    public SampleAndroidViewModelFactory() {
    }

    public SampleAndroidViewModelFactory(Application application, ISampleModel sampleModel) {
        this.application = application;
        this.sampleModel = sampleModel;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        try {
//            if (modelClass.isAssignableFrom(ISampleViewModel.class)) {
                return (T) modelClass.getConstructor(Application.class, ISampleModel.class).newInstance(application, sampleModel);
//            }
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        } catch (InstantiationException e) {
            e.printStackTrace();
        } catch (InvocationTargetException e) {
            e.printStackTrace();
        } catch (NoSuchMethodException e) {
            e.printStackTrace();
        }
        throw new IllegalArgumentException("Unknown ViewModel class");
    }
}

File: app/src/main/java/com/example/mvvm_sample/viewmodel/SampleAndroidViewModelFactory.kt
package com.example.mvvm_sample.viewmodel

import androidx.lifecycle.ViewModelProvider
import android.app.Application
import androidx.lifecycle.ViewModel
import com.example.mvvm_sample.MyApplication
import com.example.mvvm_sample.model.BaseModel
import java.lang.IllegalArgumentException
import java.lang.reflect.InvocationTargetException

class SampleAndroidViewModelFactory<R: BaseModel>(private val application: MyApplication, private val model: R) : ViewModelProvider.Factory {

    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        try {
//            if (modelClass.isAssignableFrom(ISampleViewModel.class)) {
            return modelClass.getConstructor(application::class.java, model::class.java)
                .newInstance(application, model) as T
            //            }
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        } catch (e: InstantiationException) {
            e.printStackTrace()
        } catch (e: InvocationTargetException) {
            e.printStackTrace()
        } catch (e: NoSuchMethodException) {
            e.printStackTrace()
        }
        throw IllegalArgumentException("Unknown ViewModel class")
    }
}